src/main/java/org/folio/edge/rtac/RtacHandler.java

```diff
@@ -39,76 +41,207 @@
+            rtacParams
+                .put(PARAM_FULL_PERIODICALS, Boolean.valueOf(params.get(PARAM_FULL_PERIODICALS)));
+            instanceIds = Mappers.jsonMapper.writeValueAsString(rtacParams);
+          } catch (JsonProcessingException e) {
+            logger.error("Exception during serialization in mod-rtac", e);
+            returnEmptyResponse(ctx);
             return;
           }
 
-          var rtacParams = new HashMap<>();
-          rtacParams.put(PARAM_INSTANCE_IDS, ids);
-          rtacParams.put(PARAM_FULL_PERIODICALS, Boolean.valueOf(params.get(PARAM_FULL_PERIODICALS)));
-          instanceIds = Mappers.jsonMapper.writeValueAsString(rtacParams);
-        } catch (JsonProcessingException e) {
-          logger.error("Exception during serialization in mod-rtac", e);
-          returnEmptyResponse(ctx);
-          return;
-        }
-        rtacClient.rtac(instanceIds, request.headers())
-          .thenAcceptAsync(body -> {
-            try {
-              logger.debug("rtac response: {}", body);
-              String xml;
-              final var instances = Instances.fromJson(body);
-              if (isBatch) {
-                xml = instances.toXml();
-              } else {
-                var holdings = instances.getHoldings().isEmpty() ? new Holdings() : instances.getHoldings().get(0);
-                xml = holdings.toXml();
-              }
-              logger.info("Converted Response: \n {}", xml);
-              ctx.response()
-                .setStatusCode(200)
-                .putHeader(HttpHeaders.CONTENT_TYPE, APPLICATION_XML)
-                .end(xml);
-            } catch (IOException e) {
-              logger.error("Exception translating JSON -> XML: {}", e.getMessage(), e);
-              returnEmptyResponse(ctx);
+          final boolean isAcceptHeaderExists = checkAcceptHeadersExists(request);
+          // Remember original Accept headers
+          final List<String> originAcceptHeaders = request.headers().getAll(ACCEPT);
+          final boolean acceptHeadersUpdateFlag;
+
+          if (!isAcceptHeaderExists) {
+            // Provide "json" accept header for further request to RTAC system
+            updateAcceptHeadersWith(request, Arrays.asList(JSON_MIME_TYPE));
+            acceptHeadersUpdateFlag = true;
+          } else {
+            // Replace xml-related accept header by JSON ones, to make a call to RTAC (it supports JSON by default)
+            if (checkAcceptHeadersByPattern(request, XML_RELATED_MIME_TYPE_PATTERN, false)) {
+              updateAcceptHeadersByPattern(request, XML_RELATED_MIME_TYPE_PATTERN, JSON_MIME_TYPE);
+              acceptHeadersUpdateFlag = true;
+            } else {
+              acceptHeadersUpdateFlag = false;
             }
-          })
-          .exceptionally(t -> {
-            logger.error("Exception calling mod-rtac", t);
-            returnEmptyResponse(ctx);
-            return null;
-          });
-      });
+          }
```

Comment:
Can it be extracted to a separate method? For example we can implement a `getRtacMimeType` and `updateRequestWithMimeType`?

Reply:
I agree, I'd be tempted to split this up so that the code concerned with media types and the code concerned with what information to fetch based upon whether it is a batch or not are separate if possible.

Reply:
Done.